### DISCUSSION TITLE
Improvements to settings handling

### DISCUSSION THREAD

**Discussion by vic0824** [ArcadeData/arcadedb#766] [Ideas]
**Default values**
Default values are only available in the source code and in the manual, both of which are normally not available in a production environment (which might be on a private LAN with no access to the Internet).
Every time we want to tweak some value, we have to look it up on the manual or in the source code.
I propose to put all the default values in a config file, which is read at start-up by both the Server and the Console (more on this below).

**ContextConfiguration**
I cannot think of any use case in which the difference between the GlobalConfiguration and the ContextConfiguration is relevant.
The GlobalConfiguration is already a volatile configuration, valid only for the current session (JVM), and the ContextConfiguration is derived from the GlobalConfiguration, so it has the same life-cycle.
I propose to eliminate the ContextConfiguration.

**Server settings vs Database settings**
The difference between changing a setting at JVM level and database level does not fit well with the whole range of settings, which include two distinct categories of settings:
- server settings
- database settings

Server settings must be set before the server is started, so it does not make sense to store them in the db, because by the time the db is opened, the server was already started with the default settings.
I propose to create separate mechanisms for handling server settings and database settings.
Some suggestions:
- server settings can only be specified at server level (they cannot be stored in the db) and their values are stored in a config file, which is read by the server at start-up. If the user wants to change a value, the user simply edits the config file. 
- only server settings can be specified with "-Darcadedb.<name>=<value>", or alternatively, eliminate altogether this feature, which becomes unnecessary if all server settings are read from a config file

**Selected settings available to the Console**
Some settings affect the way a new database or a new schema object is created, so they should be visible to the Console, because it is possible to create a database and its schema objects using the Console.
Other settings are only relevant for the Server, so they should not be visible to the Console.
A good starting point would be to make all database settings available to the Console.
Some settings are border-line between server and database, so a detailed analysis of the various use cases should be conducted.

**Pre-defined connection URLs for the Console**
I think it would be useful to introduce a console-specific setting containing the connection url of the default database to connect to: I have been working with ArcadeDB for 3 months and I'm already fed up of typing the full URL every time.
A more general solution would be to create a config file containing a list of aliases and their corresponding URLs, so that you can simply type "connect alias1" and the Console will connect to "remote:localhost:8080/testDb1 testUser1 testPassword1"

I realize some of these proposed changes are quite radical, but in my experience they would be beneficial in the long run, because they will spare you wasting time answering stupid (for you) questions of frustrated users who don't understand how to configure ArcadeDB.
As an example, look at how many questions about Hsqldb usage are on the SourceForge page and on StackOverflow, even though Hsqldb has a 433-pages manual: Fred Toussi spends a lot of his time answering the same questions over and over.

**Comment by gramian** [comment]
My two cents:

> Default values

* The documentation is downloadable, see: https://docs.arcadedb.com/ArcadeDB-manual.pdf solving the immediate problem.
* Having such default-values-configuration file just shifts this problem to a (IMHO much more dangerous) user responsibility. Now, as soon as this file is changed and in an isolated network, the admin has to document this changes, so future admins (including themselve) are aware that this is not a default default.

> Server settings vs Database settings

Personally, I don't want to change these settings from "within" the database, ie studio or driver, but I may want to review them to verify behaviour. So storing settings in the DB can be useful.

> Pre-defined connection URLs for the Console

I agree this can become cumbersome, I would suggest a command-line parameter or an environment variable. The latter would likely be much easier to implement, while the former would be more natural. Now if one wants to even avoid even typing addresses even once, then a wrapper script can be put around these solutions.

**Reply by vic0824** [reply]
Why do you think it is dangerous to allow the user to modify the configuration?
If you are worried that the user might set a value that renders the db unusable, I'm afraid you can't avoid it; there are already very dangerous things that the user can do:
- change a default value using a property value passed as a command-line parameter
- change a default value by issuing an SQL command using the Console
- delete the db directory
- delete jars in lib directory
- delete or modify any of the files in the config directory
- give away the root password
- modify the start-up scripts

For what concerns the objection that if you modify the default, then you lose the information of what was the original default value, there are two easy solutions:

1. download the manual, as you suggested
2. set the arcadedb.dumpConfigAtStartup property to true, so that you can look it up in the arcadedb log file

For what concerns the objection that saving server settings in the db is convenient because you might want to review them (not change them, so we are talking about default values), you are effectively saying that you would like to able to see the settings without downloading the manual, which you would be able to do if there was a config file with the default values.

Saving server settings in the db can also be misleading: you could end up with different values for the same server setting being stored in different databases, but none of them would be used, because the server needs to read the server settings before it starts (think about the settings used by Undertow), and therefore before it has the chance of reading the value stored in the db.

**Reply by gramian** [reply]
Sorry, I am not strongly objecting here, and I was only concerned with modifying a default value.

About the server setting, I was also not specific enough, sorry again, I was thinking of instances where I modified the default value and want to check if this change is (still) suitable, no matter its default.

Your scenario of misleading changed settings seems like another vote for immutable buckets https://github.com/ArcadeData/arcadedb/issues/13 😃 

PS: If you need to connect only once to a database with the console `bin/console.sh connect remote:localhost:2480/mydb user password` works already.

**Reply by vic0824** [reply]
Thanks for the tip on the Console, but I'm having problems using it in my use case.
I ship my component with a custom jre created using jlink, so I have replaced the console.sh script with my version, and it does not seem to like the connection string passed as an argument.

This is the script:
```
#!/bin/bash

DIR=$(dirname "$0")
cd "$DIR" || exit
$DIR/jre/bin/java -Dfile.encoding=UTF8 -Djava.awt.headless=true -cp "$DIR/lib/*" com.arcadedb.console.Console "connect remote:localhost:8080/csrsdb testUsr testPwd"
```

When I launch it, I get the connection message, but then the Console exits immediately:

```
[opcryosat@csrs-test csrs-orchestrator-2.0.0-SNAPSHOT]$ ./startConsole.sh
ArcadeDB Console v.23.1.1-SNAPSHOT - Copyrights (c) 2021 Arcade Data Ltd (https://arcadedb.com)
> connect remote:localhost:8080/csrsdb csrsuser CRYO_CSRS
Database 'csrsdb' connected[opcryosat@csrs-test csrs-orchestrator-2.0.0-SNAPSHOT]$ 
```

I have also tried to modify console.sh, setting $JAVA_HOME to the path of my custom jre, but it behaves in the same way as my script: it starts, it displays the connected message, but then it exits immediately.

I'm pretty sure this is all down to my poor scripting skills, so any suggestions are welcome.

**Reply by gramian** [reply]
You are aright, this is my bad: if commands are passed via commandline, then after these commands are processed the console shuts down (see: https://github.com/ArcadeData/arcadedb/blob/main/console/src/main/java/com/arcadedb/console/Console.java#L137 ).

**Reply by gramian** [reply]
@lvca This functionality makes sense, but it could also make sense to change it and make the caller manually add an `exit`.

**Reply by lvca** [reply]
It's not just the exit, but if the console finds parameters, switched to a non-interactive mode. So it won't work even by removing the exit command.

**Reply by lvca** [reply]
By the way, it works also this:

```
echo "connect mydatabase;create document type doc" | bin/console.sh
```

But it still exists after executing the commands.

I think >99% of the use cases with the console is to be interactive, so I'd change the current implementation to be interactive even when you pass commands, so when the commands are executed, it waits for more commands.

And with a new `-b` (batch mode) setting it can just execute the commands and exit.

WDYT?

**Reply by gramian** [reply]
So the cmd line args  need to be passed to console ctor and evaluated there if not empty?

**Reply by gramian** [reply]
I think it is good, it helps @vic0824 and just requires others to pass an exit at the end to keep current functionality.

**Reply by lvca** [reply]
Ok, created new issue https://github.com/ArcadeData/arcadedb/issues/767

**Reply by lvca** [reply]
Ok, it's done: https://github.com/ArcadeData/arcadedb/issues/767.

**Reply by lvca** [reply]
Now the console waits for commands (interactive mode) unless you execute an exit or you're starting the console in batch mode with `bin/console.sh -b "connect mydatabase;select from log order by date desc limit 100"`

**Comment by lvca** [comment]
I agree now there is no clear distinction between server and database settings. We could add a new property "scope" in the GlobalConfiguration enum to distinguish it, and then generate 3 tables in the docs: "JVM Settings", "Server Settings" and "Database Settings".

With this new property, we could filter the settings in the database configuration and the console, so no risks to set something not used or setting something in the wrong place.

With OrientDB we tried to make almost everybody happy, we implemented tons of features that only a few users effectively use that features. With ArcadeDB we have a totally different approach: keep the codebase as small as possible (we're still cleaning and removing code from the SQL Engine inherited from OrientDB). Having a larger codebase means harder to document, test, and maintain. Furthermore, it's harder to modify things without breaking more things. 

In ArcadeDB we're always eager to add a new cool feature, only if:
- it's something we must have
- many users request it
- a sponsor pays for the feature and at the same time it could be beneficial for others

I like there's a single point where you can configure ArcadeDB (GlobalConfiguration). Also, the fact you can store the same settings in the database is easy and useful, so you don't have to set them every time and it's portable across multiple servers.

About the server setting configuration file, I see it could be useful having a `config/server-configuration.json` that simply keeps the configuration there, as a JSON, and at startup, the server just parses the JSON and apply the settings to the GlobalConfiguration instance. But at the same time, you could just create a script that passes those settings to the server, so having a configuration file wouldn't be a game changer but it'd make things even more confusing IMO.

The URL alias is a good idea if you use the console a lot, but you can still use the console history (up/down arrows) to get previous commands and as @gramian suggests, you can connect to the server from the command line, avoiding to write URL and username all the times. So again, not a real game changer IMO.
